### PR TITLE
[FLINK-13185] [sql-parser][table-planner] Bump Calcite dependency to 1.20.0 in sql parser & flink planner

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*Bump Calcite dependency to 1.20.0 in sql parser & flink planner*

## Brief change log

  - *Upgrade version in flink-table-planner-blink/pom.xml and flink-sql-parser/pom.xml*
  - *remove indicator field from flink aggregate*
  - *fix error because SemiJoinType is deleted*
  - *update parser template file*
  - *add time functions(e.g. YEAR, MONTH) to builtin operator table*

## Verifying this change

the existing tests should pass, excluding some tests should be update due to CALCITE-2454, and parser upgrade

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by godfreyhe** [comment]
cc @danny0405 @walterddr

File: flink-table/flink-sql-parser/pom.xml [hunk 1]
```diff
@@ -51,15 +51,26 @@ under the License.
...
+				[INFO] |  +- org.apache.calcite:calcite-linq4j:jar:1.20.0:compile
+				[INFO] |  +- org.apache.commons:commons-dbcp2:jar:2.6.0:compile
+				[INFO] |  +- org.apache.commons:commons-lang3:jar:3.8:compile
+				[INFO] |  +- com.esri.geometry:esri-geometry-api:jar:2.2.0:compile
+				[INFO] |  +- com.fasterxml.jackson.core:jackson-core:jar:2.9.8:compile
+				[INFO] |  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.9.8:compile
+				[INFO] |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.9.8:compile
+				[INFO] |  +- com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:jar:2.9.8:compile
+				[INFO] |  +- com.google.guava:guava:jar:19.0:compile
+				[INFO] |  +- com.jayway.jsonpath:json-path:jar:2.4.0:compile
```
Comment (danny0405) [inline flink-table/flink-sql-parser/pom.xml]: We can exclude all these useless packages. Only `avatica-core` and `guava` are enough.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/catalog/BasicOperatorTable.scala [hunk 2]
```diff
@@ -199,6 +198,20 @@ class BasicOperatorTable extends ReflectiveSqlOperatorTable {
     ScalarSqlFunctions.REGEXP_REPLACE,
     SqlStdOperatorTable.TRUNCATE,
 
+    // TIME FUNCTIONS
+    SqlStdOperatorTable.YEAR,
+    SqlStdOperatorTable.QUARTER,
+    SqlStdOperatorTable.MONTH,
+    SqlStdOperatorTable.WEEK,
+    SqlStdOperatorTable.HOUR,
+    SqlStdOperatorTable.MINUTE,
```
Comment (danny0405) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/catalog/BasicOperatorTable.scala]: Do these added operators have their implementations now ?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/stream/sql/JoinTest.scala [hunk 3]
```diff
@@ -573,8 +575,8 @@ class JoinTest extends TableTestBase {
             term("select", "a", "b", "proctime")
           ),
           term("where",
-            "AND(=(a, a0), >=(PROCTIME(proctime), -(PROCTIME(proctime0), 3600000)), " +
-              "<=(PROCTIME(proctime), +(PROCTIME(proctime0), 3600000)))"),
+            "AND(=(a, a0), >=(PROCTIME(proctime), -(PROCTIME(proctime0), 3600000:INTERVAL HOUR))," +
+              " <=(PROCTIME(proctime), +(PROCTIME(proctime0), 3600000:INTERVAL HOUR)))"),
           term("join", "a, proctime, a0, b, proctime0"),
```
Comment (danny0405) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/stream/sql/JoinTest.scala]: nit: whitespace

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/catalog/BasicOperatorTable.scala (hunk 2 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/catalog/BasicOperatorTable.scala]: yes, there are already  several IT cases to test them.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/stream/sql/JoinTest.scala (hunk 3 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/stream/sql/JoinTest.scala]: the blank was moved from end to front, because the line exceeds limit.

File: flink-table/flink-table-planner/pom.xml [hunk 4]
```diff
@@ -148,24 +148,29 @@ under the License.
...
-				[INFO] +- org.apache.calcite:calcite-core:jar:1.18.0:compile
-				[INFO] |  +- org.apache.calcite.avatica:avatica-core:jar:1.13.0:compile
-				[INFO] |  +- org.apache.calcite:calcite-linq4j:jar:1.18.0:compile
-				[INFO] |  +- org.apache.commons:commons-lang3:jar:3.3.2:compile
-				[INFO] |  +- com.fasterxml.jackson.core:jackson-core:jar:2.9.6:compile
-				[INFO] |  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.9.6:compile
-				[INFO] |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.9.6:compile
+				"mvn dependency:tree" as of Calcite 1.20:
+
+				[INFO] +- org.apache.calcite:calcite-core:jar:1.20.0:compile
```
Comment (KurtYoung) [inline flink-table/flink-table-planner/pom.xml]: Do we have to depend on all of these? It increased a lot just for upgrading calcite version. Could you remove useless dependency?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/batch/sql/GroupWindowTest.scala [hunk 5]
```diff
@@ -336,10 +336,10 @@ class GroupWindowTest extends TableTestBase {
         ),
         term("select",
           "/(-($f0, /(*($f1, $f1), $f2)), $f2) AS EXPR$0",
-          "/(-($f0, /(*($f1, $f1), $f2)), CASE(=($f2, 1), null, -($f2, 1))) AS EXPR$1",
-          "CAST(POWER(/(-($f0, /(*($f1, $f1), $f2)), $f2), 0.5)) AS EXPR$2",
-          "CAST(POWER(/(-($f0, /(*($f1, $f1), $f2)), CASE(=($f2, 1), null, -($f2, 1))), 0.5)) " +
-            "AS EXPR$3",
+          "/(-($f0, /(*($f1, $f1), $f2)), CASE(=($f2, 1), null:BIGINT, -($f2, 1))) AS EXPR$1",
```
Comment (KurtYoung) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/batch/sql/GroupWindowTest.scala]: Did calcite change something leading on this changes? `null:BIGINT` seems weird.

File: flink-table/flink-table-planner/pom.xml (hunk 4 above)
Comment (twalthr) [inline flink-table/flink-table-planner/pom.xml]: This list should be as small as it was before.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/stream/table/OverWindowTest.scala [hunk 6]
```diff
@@ -54,20 +54,20 @@ class OverWindowTest extends TableTestBase {
             "DataStreamCalc",
             streamTableNode(table),
             // RexSimplify didn't simplify "CAST(1):BIGINT NOT NULL", see [CALCITE-2862]
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/stream/table/OverWindowTest.scala]: Remove this comment and possible other comments of this kind.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/ExpressionReductionRulesTest.scala [hunk 7]
```diff
@@ -56,16 +56,16 @@ class ExpressionReductionRulesTest extends TableTestBase {
...
+        "null:INTEGER AS EXPR$5",
+        "1990-10-24 23:00:01.123:TIMESTAMP(3) AS EXPR$6",
+        "19:BIGINT AS EXPR$7",
         "false AS EXPR$8",
         "true AS EXPR$9",
-        "2 AS EXPR$10",
+        "2:DECIMAL(2, 0) AS EXPR$10",
         "true AS EXPR$11",
-        "'trueX' AS EXPR$12"
+        "'trueX':VARCHAR(65536) CHARACTER SET \"UTF-16LE\" AS EXPR$12"
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/ExpressionReductionRulesTest.scala]: Update the test base to remove the charset. We don't need to test it. Otherwise test might be unstable depending on the environment.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/runtime/stream/table/RetractionITCase.scala [hunk 8]
```diff
@@ -110,7 +110,7 @@ class RetractionITCase extends StreamingWithStateTestBase {
       .select('count, 'count.count)
 
     val results = resultTable.toRetractStream[Row]
-    results.addSink(new StreamITCase.RetractingSink)
+    results.addSink(new StreamITCase.RetractingSink).setParallelism(1)
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/runtime/stream/table/RetractionITCase.scala]: Why is this change necessary?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/RetractionRulesTest.scala [hunk 9]
```diff
@@ -251,21 +251,17 @@ class RetractionRulesTest extends TableTestBase {
     val expected =
       unaryNode(
         "DataStreamGroupAggregate",
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/RetractionRulesTest.scala]: Could you explain this change?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/runtime/stream/table/RetractionITCase.scala (hunk 8 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/runtime/stream/table/RetractionITCase.scala]: the original flink logical plan is:
`
FlinkLogicalAggregate(group=[{0}], EXPR$0=[COUNT($0)])
  FlinkLogicalAggregate(group=[{}], EXPR$0=[SUM($0)])
    FlinkLogicalCalc(expr#0..1=[{inputs}], num=[$t1])
      FlinkLogicalDataStreamScan(id=[1], fields=[word, num])
`

after upgrade to calcite 1.20.0, the new flink logical plan is 
`
FlinkLogicalCalc(expr#0=[{inputs}], expr#1=[IS NOT NULL($t0)], expr#2=[1:BIGINT], expr#3=[0:BIGINT], expr#4=[CASE($t1, $t2, $t3)], count=[$t0], $f1=[$t4])
  FlinkLogicalAggregate(group=[{}], EXPR$0=[SUM($0)])
    FlinkLogicalCalc(expr#0..1=[{inputs}], num=[$t1])
      FlinkLogicalDataStreamScan(id=[1], fields=[word, num])
`
the change is cause by [CALCITE-896](http://issues.apache.org/jira/browse/CALCITE-896) in 1.20.0.

the physical plan is
`
DataStreamCalc(select=[EXPR$0 AS count, CASE(IS NOT NULL(EXPR$0), 1:BIGINT, 0:BIGINT) AS $f1])
  DataStreamGroupAggregate(select=[SUM(num) AS EXPR$0])
    DataStreamCalc(select=[num])
      DataStreamScan(id=[1], fields=[word, num])
`

the parallelism of transformations corresponding to the top `DataStreamCalc` and `DataStreamGroupAggregate` is 1, while the parallelism of sink transformation is 4 (default value), so we should set the parallelism of sink as 1 to avoid shuffling data.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/batch/sql/GroupWindowTest.scala (hunk 5 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/batch/sql/GroupWindowTest.scala]: due to https://issues.apache.org/jira/browse/CALCITE-2454

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/RetractionRulesTest.scala (hunk 9 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/RetractionRulesTest.scala]: before upgrade, the flink logical plan is 
```
FlinkLogicalAggregate(group=[{0}], EXPR$0=[COUNT($0)])
  FlinkLogicalUnion(all=[true])
    FlinkLogicalCalc(expr#0..1=[{inputs}], count=[$t1])
      FlinkLogicalAggregate(group=[{0}], EXPR$0=[COUNT($1)])
        FlinkLogicalDataStreamScan(id=[1], fields=[word, number])
    FlinkLogicalCalc(expr#0..1=[{inputs}], count=[$t1])
      FlinkLogicalDataStreamScan(id=[2], fields=[word_r, count_r])
```

after upgrade, the flink logical plan is
```
FlinkLogicalAggregate(group=[{0}], EXPR$0=[COUNT($0)])
  FlinkLogicalCalc(expr#0..1=[{inputs}], count=[$t1])
    FlinkLogicalUnion(all=[true])
      FlinkLogicalAggregate(group=[{0}], EXPR$0=[COUNT($1)])
        FlinkLogicalDataStreamScan(id=[1], fields=[word, number])
      FlinkLogicalDataStreamScan(id=[2], fields=[word_r, count_r])
```

however, their costs are totally same on the second-top level node
```
	rel#99:Subset#6.LOGICAL, best=rel#107, importance=0.9
		rel#107:FlinkLogicalCalc.LOGICAL(input=RelSubset#102,expr#0..1={inputs},count=$t1), rowcount=110.0, cumulative cost={520.0 rows, 410.0 cpu, 5200.0 io}
		rel#118:FlinkLogicalUnion.LOGICAL(input#0=RelSubset#116,input#1=RelSubset#117,all=true), rowcount=110.0, cumulative cost={520.0 rows, 410.0 cpu, 5200.0 io}
```
some behaviors are changed (maybe [CALCITE-2897](https://issues.apache.org/jira/browse/CALCITE-2897), i do not have time to debug the details) in volcano planner which cause different choice.

**Comment by KurtYoung** [comment]
python tests are failed with: 
E                   Caused by: java.lang.Exception: java.lang.NoClassDefFoundError: org/apache/commons/codec/language/Soundex

**Comment by KurtYoung** [comment]
The reason might be:
				<exclusion>
					<groupId>commons-codec</groupId>
					<artifactId>commons-codec</artifactId>
				</exclusion>

I think blink planner will also suffer this issue, it's not failed because python table API didn't run with blink planner

**Comment by godfreyhe** [comment]
Oops.. I only had validated the dependency in flink-table module. I will re-validate it in whole project

**Comment by KurtYoung** [comment]
> Oops.. I only had validated the dependency in flink-table module. I will re-validate it in whole project

Maybe also find a way to verify python table api with blink planner.

**Comment by godfreyhe** [comment]
> > Oops.. I only had validated the dependency in flink-table module. I will re-validate it in whole project
> 
> Maybe also find a way to verify python table api with blink planner.

ok, i will fix it

**Comment by godfreyhe** [comment]
> > > Oops.. I only had validated the dependency in flink-table module. I will re-validate it in whole project
> > 
> > 
> > Maybe also find a way to verify python table api with blink planner.
> 
> ok, i will fix it

@dianfu told me that flink-python does not support blink planner now

**Comment by KurtYoung** [comment]
Seem we have some irrelevant fails:
https://travis-ci.org/KurtYoung/flink/jobs/557153197
https://travis-ci.org/godfreyhe/flink/builds/557150193

Merging this..

**Comment by tillrohrmann** [comment]
Hi guys, I think this PR changes Flink's dependencies. Whenever we change dependencies we need to update the corresponding NOTICE files. Please correct this and make sure that future PRs which contain these kind of changes also update the NOTICE files @KurtYoung.

See https://cwiki.apache.org/confluence/display/FLINK/Licensing for more information about licensing.

**Comment by KurtYoung** [comment]
> Hi guys, I think this PR changes Flink's dependencies. Whenever we change dependencies we need to update the corresponding NOTICE files. Please correct this and make sure that future PRs which contain these kind of changes also update the NOTICE files @KurtYoung.
> 
> See https://cwiki.apache.org/confluence/display/FLINK/Licensing for more information about licensing.

Oops, I forgot this, will add it now.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/CommonCorrelate.scala [hunk 10]
```diff
@@ -46,7 +47,7 @@ trait CommonCorrelate {
     inputSchema: RowSchema,
     udtfTypeInfo: TypeInformation[Any],
     returnSchema: RowSchema,
-    joinType: SemiJoinType,
+    joinType: JoinRelType,
```
Comment (walterddr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/CommonCorrelate.scala]: I notice that this change added a number of new warnings regarding that the new `ANIT` and `SEMI` joins are not handled in some of the switch case clause across the table planner. Are we planning to have a follow up JIRA to fix this?

Comment (godfreyhe) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/CommonCorrelate.scala]: sorry for late reply, I create a JIRA to fix this: https://issues.apache.org/jira/browse/FLINK-13347
